src/core/namespace.c

```diff
@@ -1173,19 +1173,15 @@ int setup_namespace(
 
         if (root_directory)
                 root = root_directory;
-        else if (root_image || n_bind_mounts > 0 || n_temporary_filesystems > 0) {
-
-                /* If we are booting from an image, create a mount point for the image, if it's still missing. We use
-                 * the same mount point for all images, which is safe, since they all live in their own namespaces
-                 * after all, and hence won't see each other. We also use such a root directory whenever there are bind
-                 * mounts configured, so that their source mounts are never obstructed by mounts we already applied
-                 * while we are applying them. */
+        else {
+                /* Always create the mount namespace in a temporary directory, instead of operating
+                 * directly in the root. The temporary directory prevents any mounts from being
+                 * potentially obscured my other mounts we already applied. */
```

Comment:
the part "We use the same mount point for all images, which is safe, since they all live in their own namespacesafter all, and hence won't see each other. " really should stay. I mean, it's not obvious that it is sufficient to use the same directory for setting this up here, and this deserves explicit mention